File: compiler/frontend/ast/expr.cpp
#include "compiler/frontend/ast/expr.hpp"

#include <algorithm>

namespace frontend::ast {
lex::Span UnaryExpr::span() const {
    auto b = op.span.begin;
    auto e = value->span().end;
    return lex::Span{b, e};
}

lex::Span BinaryExpr::span() const {
    auto b = lhs->span().begin;
    auto e = rhs->span().end;
    return lex::Span{b, e};
}

lex::Span NumLitExpr::span() const {
    return value.span;
}

lex::Span BoolLitExpr::span() const {
    return value.span;
}

lex::Span VarRefExpr::span() const {
    return name.span;
}

lex::Span MemberRefExpr::span() const {
    auto b = target->span().begin;
    auto e = member.span.end;
    return lex::Span{b, e};
}

lex::Span CallExpr::span() const {
    auto calleeSpan = callee->span();
    size_t e;
    if (args.empty()) {
        e = calleeSpan.end;
    } else {
        e = args.back()->span().end;
    }

    return lex::Span{calleeSpan.begin, e};
}

lex::Span StructInitExpr::span() const {
    size_t end = 0;

    for (const auto& [_, field] : fields) {
        end = std::max(end, field->span().end);
    }

    return lex::Span{tySpan.begin, end};
}
}  // namespace frontend::ast

File: compiler-fuzzing/src/main/scala/Expr.scala
import scala.collection.Seq
import scala.math.Ordering.BigIntOrdering

sealed trait Expr

case class UnaryMinusExpr(value: Expr) extends Expr {
  override def toString: String = s"-($value)"
}

case class NotExpr(value: Expr) extends Expr {
  override def toString: String = s"!($value)"
}

sealed trait BinaryExpr(op: String) extends Expr {
  def lhs: Expr
  def rhs: Expr
  override def toString: String = s"($lhs) $op ($rhs)"
}

case class EqExpr(lhs: Expr, rhs: Expr) extends BinaryExpr("==")
case class NeExpr(lhs: Expr, rhs: Expr) extends BinaryExpr("!=")

case class LtExpr(lhs: Expr, rhs: Expr) extends BinaryExpr("<")
case class GtExpr(lhs: Expr, rhs: Expr) extends BinaryExpr(">")
case class LeExpr(lhs: Expr, rhs: Expr) extends BinaryExpr("<=")
case class GeExpr(lhs: Expr, rhs: Expr) extends BinaryExpr(">=")

case class AddExpr(lhs: Expr, rhs: Expr) extends BinaryExpr("+")
case class SubExpr(lhs: Expr, rhs: Expr) extends BinaryExpr("-")
case class MulExpr(lhs: Expr, rhs: Expr) extends BinaryExpr("*")
case class DivExpr(lhs: Expr, rhs: Expr) extends BinaryExpr("/")

case class OrExpr(lhs: Expr, rhs: Expr)  extends BinaryExpr("||")
case class AndExpr(lhs: Expr, rhs: Expr) extends BinaryExpr("&&")

case class NumLitExpr(value: Int) extends Expr {
  override def toString: String = value.toString
}
case class BoolLitExpr(value: Boolean) extends Expr {
  override def toString: String = value.toString
}

case object NullLitExpr extends Expr {
  override def toString: String = "null"
}
case class VarRefExpr(name: String) extends Expr {
  override def toString: String = name
}
case class MemberRefExpr(target: Expr, member: String) extends Expr {
  override def toString: String = s"($target).$member"
}
case class CallExpr(callee: Expr, args: Seq[Expr]) extends Expr {
  override def toString: String = s"($callee)" ++ args.mkString("(", ", ", ")")
}
case class StructInitExpr(ty: StructTy, fields: Seq[(String, Expr)]) extends Expr {
  override def toString: String =
    s"${ty.name}" ++ fields.map { (name, value) => s".$name = $value" }.mkString("{", ",\n", "}")
}
